awx/main/scheduler/dag_simple.py

```diff
@@ -27,98 +59,169 @@ def __len__(self):
+            nodes = []
+            map(lambda l: nodes.extend(self._get_dependencies_by_label(this_ord, l)),
+                self.node_from_edges_by_label.keys())
+            return nodes
+
+    def _get_dependents_by_label(self, node_index, label):
+        return [self.nodes[index] for index in
+                self.node_to_edges_by_label.get(label, {})
+                                           .get(node_index, [])]
 
     def get_dependents(self, obj, label=None):
-        decendents = []
         this_ord = self.find_ord(obj)
-        for node, dep, lbl in self.edges:
-            if label:
-                if dep == this_ord and lbl == label:
-                    decendents.append(self.nodes[node])
+        nodes = []
+        if label:
+            return self._get_dependents_by_label(this_ord, label)
+        else:
+            nodes = []
+            map(lambda l: nodes.extend(self._get_dependents_by_label(this_ord, l)),
+                self.node_to_edges_by_label.keys())
+            return nodes
+
+    def get_root_nodes(self):
+        return [self.nodes[index] for index in self.root_nodes]
+
+    def has_cycle(self):
+        node_objs = [node['node_object'] for node in self.get_root_nodes()]
+        node_objs_visited = set([])
+        path = set([])
+        stack = node_objs
+        res = False
+
+        if len(self.nodes) != 0 and len(node_objs) == 0:
+            return True
+
+        while stack:
+            node_obj = stack.pop()
+
+            children = [node['node_object'] for node in self.get_dependencies(node_obj)]
+            children_to_add = filter(lambda node_obj: node_obj not in node_objs_visited, children)
+
+            if children_to_add:
+                if node_obj in path:
+                    res = True
+                    break
+                path.add(node_obj)
+                stack.append(node_obj)
+                stack.extend(children_to_add)
             else:
-                if dep == this_ord:
-                    decendents.append(self.nodes[node])
-        return decendents
+                node_objs_visited.add(node_obj)
+                path.discard(node_obj)
+        return res
 
-    def get_leaf_nodes(self):
-        leafs = []
-        for n in self.nodes:
-            if len(self.get_dependencies(n['node_object'])) < 1:
-                leafs.append(n)
-        return leafs
+    def sort_nodes_topological(self):
+        nodes_sorted = deque()
+        obj_ids_processed = set([])
 
-    def get_root_nodes(self):
-        roots = []
-        for n in self.nodes:
-            if len(self.get_dependents(n['node_object'])) < 1:
-                roots.append(n)
-        return roots
+        def visit(node):
+            obj = node['node_object']
+            if obj.id in obj_ids_processed:
+                return
+
+            for child in self.get_dependencies(obj):
+                visit(child)
+            obj_ids_processed.add(obj.id)
+            nodes_sorted.appendleft(node)
```

Comment:
so dependencies are parents.

the node is appended to the left of the sorted list.

the node is added to the left of the list after its parents are.

Thus, the greatest depth nodes come first?